packages/opentelemetry-exporter-collector/src/CollectorMetricExporterBase.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { MetricExporter, MetricRecord } from '@opentelemetry/metrics';
+import { Attributes, Logger } from '@opentelemetry/api';
+import { CollectorExporterConfigBase } from './types';
+import { NoopLogger, ExportResult } from '@opentelemetry/core';
+import * as collectorTypes from './types';
+
+const DEFAULT_SERVICE_NAME = 'collector-metric-exporter';
+
+/**
+ * Collector Metric Exporter abstract base class
+ */
+export abstract class CollectorMetricExporterBase<
+  T extends CollectorExporterConfigBase
+> implements MetricExporter {
+  public readonly serviceName: string;
+  public readonly url: string;
+  public readonly logger: Logger;
+  public readonly hostName: string | undefined;
+  public readonly attributes?: Attributes;
+  protected readonly _startTime = new Date().getTime() * 1000000;
+  private _isShutdown: boolean = false;
+
+  /**
+   * @param config
+   */
+  constructor(config: T = {} as T) {
+    this.logger = config.logger || new NoopLogger();
+    this.serviceName = config.serviceName || DEFAULT_SERVICE_NAME;
+    this.url = this.getDefaultUrl(config.url);
+    this.attributes = config.attributes;
+    if (typeof config.hostName === 'string') {
+      this.hostName = config.hostName;
+    }
+    this.onInit();
+  }
+
+  /**
+   * Export metrics
+   * @param metrics
+   * @param resultCallback
+   */
+  export(
+    metrics: MetricRecord[],
+    resultCallback: (result: ExportResult) => void
+  ) {
+    if (this._isShutdown) {
+      resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+      return;
+    }
+
+    this._exportMetrics(metrics)
+      .then(() => {
+        resultCallback(ExportResult.SUCCESS);
+      })
+      .catch(
+        (
+          error: collectorTypes.opentelemetryProto.collector.trace.v1.ExportTraceServiceError
```

Comment:
This looks fishy, why are we using `ExportTraceServiceError` for metric?

Reply:
Ah yeah, that type isn't actually specified in the proto. It's just a template for errors. So, I'm renaming it to `ExportServiceError` and moving it out of the proto